core/src/main/java/org/axonframework/queryhandling/SimpleQueryBus.java

```diff
@@ -228,103 +227,48 @@ private boolean unsubscribe(String queryName,
-    }
-
-    @Override
-    public Map<SubscriptionQueryMessage<?, ?, ?>, Long> requestedFromDownstream(
-            Predicate<SubscriptionQueryMessage<?, ?, ?>> filter) {
-        return updateHandlers.keySet()
-                             .stream()
-                             .filter(filter)
-                             .collect(Collectors.toMap(Function.identity(),
-                                                       m -> updateHandlers.get(m).requestedFromDownstream()));
-    }
-
-    private <Q, I, U> void initialResult(SubscriptionQueryMessage<Q, I, U> query,
-                                         MessageMonitor.MonitorCallback monitorCallback, MonoSinkWrapper<QueryResponseMessage<I>> monoSink) {
-        Iterator<MessageHandler<? super QueryMessage<?, ?>>> handlerIterator = getHandlersForMessage(query).iterator();
-        boolean foundHandler = false;
-        while (!foundHandler && handlerIterator.hasNext()) {
-            DefaultUnitOfWork<QueryMessage<Q, I>> uow = DefaultUnitOfWork.startAndGet(query);
-            try {
-                interceptAndInvoke(uow, handlerIterator.next())
-                        .thenAccept(initialResponse -> {
-                            monoSink.success(initialResponse);
-                            monitorCallback.reportSuccess();
-                        })
-                        .exceptionally(error -> {
-                            monoSink.error(error);
-                            monitorCallback.reportFailure(error);
-                            return null;
-                        });
-                foundHandler = true;
-            } catch (NoHandlerForQueryException e) {
-                // Ignore this Query Handler, as we may have another one which is suitable
-            } catch (Exception e) {
-                foundHandler = true;
-                monitorCallback.reportFailure(e);
-                monoSink.error(e);
-                break;
-            }
-        }
-        if (!foundHandler) {
-            NoHandlerForQueryException error = new NoHandlerForQueryException(
-                    format("No suitable handler was found for %s with response type %s and update type %s",
-                           query.getQueryName(),
-                           query.getResponseType(),
-                           query.getUpdateResponseType()));
-            monitorCallback.reportFailure(error);
-            monoSink.error(error);
-        }
-    }
-
-    private <U> void updates(SubscriptionQueryMessage<?, ?, U> query,
-                             FluxSinkWrapper<SubscriptionQueryUpdateMessage<U>> fluxSink) {
-        updateHandlers.put(query, fluxSink);
-        fluxSink.onDispose(() -> updateHandlers.remove(query));
-        Optional.ofNullable(delayedUpdates.remove(query))
-                .ifPresent(delays -> delays.forEach(c -> c.accept(query, fluxSink)));
+                      .forEach(query -> updateHandlers.remove(query).forEach(uh -> uh.error(cause)));
```

Comment:
same here: this could be a concurrency issue if the handlers are closed concurrently (`remove` will return `null`)